lib/_http_server.js

```diff
@@ -930,7 +934,7 @@ function parserOnIncoming(server, socket, state, req, keepAlive) {
 }
 
 function resetSocketTimeout(server, socket, state) {
-  if (!state.keepAliveTimeoutSet)
+  if (!state.keepAliveTimeoutSet || state.responseSent)
```

Comment:
I think you can use `!state._dumped` instead: https://github.com/nodejs/node/blob/a8a86387fa4a8bb2c0e058c5c16f3fa666f15fe7/lib/_http_incoming.js#L92-L94

Reply:
Sounds nice ... but `_dumped` is not a property of `state` but of IncomingMessage (req), right?

Would it be correct to use  something like `state.incoming[0]._dumped` at this time? But `state.incoming` is already shifted in `resOnFinish()` ...

https://github.com/nodejs/node/blob/a8a86387fa4a8bb2c0e058c5c16f3fa666f15fe7/lib/_http_server.js#L800

Reply:
@ronag do you think we can use one of the other properties we have?

Reply:
Isn't it enough to just check `state.incoming.length > 0`?

Reply:
i.e. keepAlive timeout should only run if `state.incoming.length === 0` and only be reset if `state.incoming.length > 0`.

Reply:
Oh yes, that's a good idea ... I'll try tomorrow. Thanks!

Reply:
Ok, resetting the timeout only if state.outgoing.length > 0 (or rather "return in resetSocketTimeout() if outgoing.length === 0") works (I force pushed the new commit). But also breaks two other tests, which tests "slow requests" - in detail, when the request-headers do not fully arrive in keep-alive-timeout, it fails:

- sequential/test-http-server-keep-alive-timeout-slow-client-headers.js
- sequential/test-http-server-keep-alive-timeout-slow-server.js

So the condition for resetting the keep-alive-timeout has to take into account, if a new request also has started arriving, I think. I currently do not have a full overview over the HTTP server state machine.

Or maybe it is better, to not call `resetSocketTimeout()` every time when socket-data arrives, but really only on the first byte of a new request? Any further ideas?

Reply:
How about `state.outgoing.length && state.incoming.length`?

Reply:
I tested with this code in `resetSocketTimeout()` ... so "don't reset timeout if outgoing or incoming packets are in the queue":

```javascript
if (!state.keepAliveTimeoutSet)
  return;

if (state.outgoing.length === 0 && state.incoming.length === 0)
  return;
```

... but it does not change anything (fixes this issue, still breaks e.g. "sequential/test-http-server-keep-alive-timeout-slow-client-headers.js"). It seems that `state.incoming` is filled after request-headers are fully received:

`state.incoming` is filled in `parserOnIncoming()`:
https://github.com/nodejs/node/blob/a8a86387fa4a8bb2c0e058c5c16f3fa666f15fe7/lib/_http_server.js#L861
... which seems to be called on `parserOnHeadersComplete()`:
https://github.com/nodejs/node/blob/a8a86387fa4a8bb2c0e058c5c16f3fa666f15fe7/lib/_http_common.js#L127

When logging `state.incoming.length` in `resetSocketTimeout()` it is 0 also for the failing "test-http-server-keep-alive-timeout-slow-client-headers".

Do we have any flag/info, if we already received the first byte of the next request? Or is it hidden by `HTTPParser`? At least `parserOnHeaders()` seems to be interesting ... but is seemingly not exposed:

https://github.com/nodejs/node/blob/a8a86387fa4a8bb2c0e058c5c16f3fa666f15fe7/lib/_http_common.js#L63

Reply:
I guess we should use `parser[kOnMessageBegin]` to set some kind of flag that gets reset in `parser[kOnHeadersComplete]`.

Reply:
Or maybe just move the `resetSocketTimeout()` call from `onParserExecuteCommon()` to a (new) callback behind `parser[kOnMessageBegin]`? IMHO there is no other situation where the keep-aliv-timeout shoud be reset? Maybe we can even get rid of the `keepAliveTimeoutSet` flag then? Or do I miss anything?

Reply:
SGTM

Reply:
Ok, I'll try later :-)

Reply:
I'm still stuck here ... it seems that `parserOnHeaders()` in _http_common.js (`HTTPParser.kOnHeaders` native callback) is never called, even for incomplete headers received (but maybe I made a mistake while debugging).

I also tried the `HTTPParser.kOnMessageBegin` native callback (stolen here https://github.com/nodejs/node/blob/master/src/node_http_parser.cc#L72) ... which seems also not to be called in all my testcases.

I currently do not see a clean solution, without understanding/changing the native code and "llhttp" (And I understand why the timeout-handling is implemented that way currently in NodeJS)